composer/src/lib.rs

```diff
@@ -159,49 +159,87 @@ pub struct ComposerPackage {
+                formatter.write_str("an array of repository definitions (or e.g. '{\"packagist.org\": false}' to disable a repo), or an object of repository names as keys and repository definitions (or boolean false to disable a repo) as values")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let mut values: Vec<ComposerRepository> =
+                    Vec::with_capacity(seq.size_hint().unwrap_or(0));
+                let mut index = 0;
+                while let Some(value) = seq.next_element::<Value>()? {
+                    match serde_json::from_value::<ComposerRepository>(value.clone()) {
+                        Ok(r) => values.push(r),
+                        Err(_) => {
+                            match serde_json::from_value::<ComposerRepositoryDisablement>(value) {
+                                Ok(d) => values.push(ComposerRepository::Disabled(d)),
+                                Err(_) => {
+                                    return Err(Error::custom(format!("invalid value at array index {index}, expected repository definition object or disablement using single-key object notation (e.g. `{{\"packagist.org\": false}}`)")))
+                                },
+                            }
+                        }
+                    }
+                    index += 1;
+                }
+                Ok(ComposerRepositories(values))
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: MapAccess<'de>,
+            {
+                let mut values = Vec::with_capacity(map.size_hint().unwrap_or(0));
+                // we fetch entries as serde_json::Value objects so that we can handle the disabled repo (key: $name, value: false) case
+                while let Some((key, value)) = map.next_entry::<String, Value>()? {
+                    match serde_json::from_value::<ComposerRepository>(value.clone()) {
+                        // it de-serialized fine
+                        Ok(r) => values.push(r),
+                        Err(_) => {
+                            // try de-serializing as a boolean false
+                            match serde_json::from_value::<MustBe!(false)>(value) {
+                                // that's it; we make a "Disabled" repo variant using the key as the name
+                                Ok(_) => values.push(ComposerRepository::Disabled(ComposerRepositoryDisablement::new(&key))),
+                                Err(_) => {
+                                    return Err(Error::custom(format!("invalid value at object key '{key}', expected repository definition object or disablement using boolean `false`")))
+                                },
+                            }
+                        }
+                    }
```

Comment:
Optional suggestion. Remove duplicate push call with functional style code (+8 -12)

```suggestion
                    let repo =
                        serde_json::from_value::<ComposerRepository>(value.clone()).or_else(|_| {
                            // try de-serializing as a boolean false
                            serde_json::from_value::<MustBe!(false)>(value)
                                // that's it; we make a "Disabled" repo variant using the key as the name
                                .map(|_| ComposerRepository::Disabled(ComposerRepositoryDisablement::new(&key)))
                        })
                        .map_err(|_| Error::custom(format!("invalid value at object key '{key}', expected repository definition object or disablement using boolean `false`")))
                        ?;
                    values.push(repo);
```